lib/mayaUsd/ufe/UsdUndoUngroupCommand.cpp

```diff
@@ -0,0 +1,117 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "UsdUndoUngroupCommand.h"
+
+#include "private/UfeNotifGuard.h"
+
+#include <mayaUsd/ufe/Utils.h>
+#include <mayaUsd/undo/UsdUndoBlock.h>
+#include <mayaUsdUtils/util.h>
+
+#include <pxr/usd/sdf/changeBlock.h>
+#include <pxr/usd/usd/editContext.h>
+
+#include <ufe/globalSelection.h>
+#include <ufe/hierarchy.h>
+#include <ufe/observableSelection.h>
+#include <ufe/scene.h>
+
+PXR_NAMESPACE_USING_DIRECTIVE
+
+namespace MAYAUSD_NS_DEF {
+namespace ufe {
+
+UsdUndoUngroupCommand::UsdUndoUngroupCommand(const UsdSceneItem::Ptr& groupItem)
+    : Ufe::UndoableCommand()
+    , _groupItem(groupItem)
+    , _compositeInsertCmd(std::make_shared<Ufe::CompositeUndoableCommand>())
+{
+}
+
+UsdUndoUngroupCommand::~UsdUndoUngroupCommand() { }
+
+UsdUndoUngroupCommand::Ptr UsdUndoUngroupCommand::create(const UsdSceneItem::Ptr& groupItem)
+{
+    return std::make_shared<UsdUndoUngroupCommand>(groupItem);
+}
+
+void UsdUndoUngroupCommand::execute()
+{
+    // "Ungrouping" means moving group's children up a level in hierarchy
+    // followed by group node getting removed.
+
+    // move group's children one level up
+    auto groupHier = Ufe::Hierarchy::hierarchy(_groupItem);
+    auto levelUpHier = Ufe::Hierarchy::hierarchy(groupHier->parent());
+    for (auto& child : groupHier->children()) {
+        auto insertChildCmd = levelUpHier->appendChildCmd(child);
+        _compositeInsertCmd->append(insertChildCmd);
+    }
+
+    _compositeInsertCmd->execute();
+
+    // remove group prim
+    MayaUsd::ufe::InAddOrDeleteOperation ad;
+    UsdUndoBlock                         undoBlock(&_undoableItem);
+
+    auto           stage = _groupItem->prim().GetStage();
+    UsdEditContext ctx(stage, stage->GetEditTarget().GetLayer());
+    auto           retVal = stage->RemovePrim(_groupItem->prim().GetPath());
+    TF_VERIFY(retVal, "Failed to remove '%s'", _groupItem->prim().GetPath().GetText());
+}
+
+void UsdUndoUngroupCommand::undo()
+{
+    MayaUsd::ufe::InAddOrDeleteOperation ad;
+
+    _undoableItem.undo();
+
+    // After undo, the group prim loses it's specifier and type.
+    // At this point _groupItem->prim() is stale and should not be used.
+    // Define a new prim from ufe path to get back the original concrete xfrom prim.
+    auto stage = getStage(_groupItem->path());
+    TF_AXIOM(stage);
+
+    auto groupUsdPath = getPath(_groupItem->path());
+    auto groupPrim = stage->DefinePrim(groupUsdPath, TfToken("Xform"));
```

Comment:
On thinking about this one a bit more, this line has me worried.  What if the prim we removed is not of type "Xform", but rather some funky plugin type?  What if it had custom metadata, or animations, or property values?  E.g. what if the original grouping Xform had an animated transform?  Here, you're just bringing back a brand new, vanilla Xform, so on undo, the prim won't be brought back to its original state.  I hope there's a way to use the UsdStateDelegate to bring back the full state of the original prim.

Reply:
Hmmm... I was under the impression that the UsdStateDelegate would bring back the properties but now that is not the case after doing some tests.  That's a shame....

Ok, I need to think about this.  If somehow I can store the property specs for the removed prim on execute(), then I can transfer it back to the new defined prim on undo(). Something similar to SdfCopySpec.